teps/0113-produce-results-from-matrixed-pipelinetasks.md

```diff
@@ -0,0 +1,204 @@
+want to pass them to subsequent `PipelineTasks` that to process the test outputs.
+
+## Background
+
+This section describes the using `Results` in `Pipelines` - see [docs][results-docs] for further details.
+
+### String Results
+
+String `Results` are is a beta feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>)`.
+
+String `Results` from previous `TaskRuns` can be passed into the `Matrix` can be passed individually into the `Matrix`:
+
+```yaml
+tasks:
+...
+- name: task-4
+  taskRef:
+    name: task-4
+  matrix:
+  - name: values
+    value: 
+    - (tasks.task-1.results.foo) # string
+    - (tasks.task-2.results.bar) # string
+    - (tasks.task-3.results.rad) # string
+```
+
+### Array Results
+
+Array `Results` is an alpha feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>[*])`.
+Array indexing is supported and is referred as `$(tasks.<pipelinetask-name>.results.<result-name>[i])` where `i`
+is the index.
+
+Array `Results` from previous `TaskRuns` can be passed into the `Matrix` can be passed individually into the `Matrix`:
+
+```yaml
+tasks:
+...
+- name: task-5
+  taskRef:
+    name: task-5
+  matrix:
+  - name: values
+    value: (tasks.task-4.results.foo[*]) # array
+```
+
+For further information, see [TEP-0076: Array Results][tep-0076]. 
+
+### Object Results
+
+Object `Results` is an as alpha feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>[*])`. 
+Object elements can be referred to as `$(tasks.<pipelinetask-name>.results.<result-name>.key)`.
+
+For further information, see [TEP-0075: Object Parameters and Results][tep-0075].
+
+## Proposal
+
+`Matrix` in a `PipelineTask` take `Parameters` of type `Array` to replace `Parameters` of type `String` in the `Task`. 
+We propose that `PipelineTasks` with `Matrix` produce `Results` of type `Array` that aggregate `Results` of type 
+`String` in the underlying `Tasks`.
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pr
+spec:
+  params:
+  - name: foo
+    value: 
+      - foo-1
+      - foo-2
+      - foo-3
+  pipelineSpec:
+    params:
+    - name: foo
+      type: array
+    tasks:
+    - name: pipeline-task-1 # will create 3 taskruns
+      matrix:
+      - name: foo
+        value: $(params.foo) # array parameter with 3 items   
+      taskSpec:
+        params:
+        - name: foo
+          type: string
+        results:
+        - name: bar
+          type: string          
+        steps:
+          ...
+    - name: pipeline-task-2
+      param:
+      - name: bars
+        value: $(tasks.pipeline-task-1.results.bar[*]) # array result with 3 items
+      taskSpec:
+        params:
+        - name: bars
+          type: string
+        steps:
+          ...       
+```
+
+### Notes
+
+#### Array and Object Results in underlying Tasks
+
+In `Matrix` we don't support replacing `Parameters` of type `Array` or `Object` in the underlying `Task` because we
```

Comment:
to check my understanding, what you are saying here is that you can have:
```
    matrix:
    - name: IMAGE
      value: $(params.images)
```
where images is an array of strings, and each string is fanned out, but you couldn't have
```
    matrix:
    - name: array
      value: $(params.2d-array)
```
where 2d-array is an array of arrays, and each matrixed pipelinetask receives a single array param? and likewise you can't have
```
    matrix:
    - name: object
      value: $(params.array-of-object)
```
where each matrixed pipelinetask receives its own object param. Is that right?

Reply:
exactly, and that's because we don't have `2d-array` or `array-of-objects` yet - this is something we can explore in future work